### PR TITLE
feat: add endian aware convert functions

### PR DESCRIPTION
**Pull request by anonrig**
The last release broke the Node.js builds. Upon request from @addaleax, I'm adding new function that handles the endianness.

Fixes: https://github.com/nodejs/node/pull/46208

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -770,6 +784,20 @@ They are well suited when you expect that there might be errors in the input tha
...
+ *
+ * During the conversion also validation of the input string is done.
+ * This function is suitable to work with inputs from untrusted sources.
+ *
+ * @param input         the UTF-8 string to convert
+ * @param length        the length of the string in bytes
+ * @param utf16_buffer  the pointer to buffer that can hold conversion result
+ * @return a result pair struct with an error code and either the position of the error if any or the number of char16_t written if successful.
+ */
+simdutf_warn_unused result convert_utf8_to_utf16_with_errors(const char * input, size_t length, char16_t* utf16_output) noexcept;
```
Comment (addaleax) [inline README.md:799]: There's a lot more functions that currently use `utf16le` and `utf16be`, not just the UTF8-to-UTF16 conversion functions

**Comment by lemire** [comment]
The tests at https://github.com/simdutf/simdutf/blob/master/tests/bele_tests.cpp defines the expected behaviour.

**Comment by lemire** [comment]
@anonrig Don't implement the methods in 'implementation' as will take you too much work.

**Comment by lemire** [comment]
You only need to add your function into the first section of...
https://github.com/simdutf/simdutf/blob/master/include/simdutf/implementation.h
and at the end of...
https://github.com/simdutf/simdutf/blob/master/src/implementation.cpp


Don't change anything in `implementation::` otherwise you will just work way harder than needed.

**Comment by lemire** [comment]
BTW the idea is fine... I am just worried that you are going to write way more code than you need to. It is actually a fairly quick hack. No need to go deep.

**Comment by lemire** [comment]
@anonrig Call me on signal if you don't understand what I am writing.

Cheers!!!

File: src/implementation.cpp [hunk 2]
```diff
@@ -687,36 +743,78 @@ simdutf_warn_unused result convert_utf32_to_utf8_with_errors(const char32_t * bu
...
+  return active_implementation->convert_utf32_to_utf16le(buf, len, utf16_buffer);
+  #endif
+}
 simdutf_warn_unused size_t convert_utf32_to_utf16le(const char32_t * buf, size_t len, char16_t* utf16_buffer) noexcept {
   return active_implementation->convert_utf32_to_utf16le(buf, len, utf16_buffer);
 }
 simdutf_warn_unused size_t convert_utf32_to_utf16be(const char32_t * buf, size_t len, char16_t* utf16_buffer) noexcept {
   return active_implementation->convert_utf32_to_utf16be(buf, len, utf16_buffer);
 }
+simdutf_warn_unused result convert_utf32_to_utf16_with_errors(const char32_t * buf, size_t len, char16_t* utf16_buffer) noexcept {
```
Comment (lemire) [inline src/implementation.cpp:759]: This construction is fine but could be even simpler!!!
```C++
 #if SIMDUTF_IS_BIG_ENDIAN
   return convert_utf32_to_utf16be_with_errors(buf, len, utf16_buffer);
   #else
   return convert_utf32_to_utf16le_with_errors(buf, len, utf16_buffer);
   #endif
```

Comment (lemire) [reply src/implementation.cpp:759]: (It is not a request to change your code.)

Comment (anonrig) [reply src/implementation.cpp:759]: You're right. I'll change them.

**Comment by lemire** [comment]
@anonrig I think that @addaleax is right... throughout, maybe you could replace, e.g.,

```
Without depending on the system endianness;
    * Compute the number of bytes that this UTF-16LE 
```

with

```
    * Compute the number of bytes that this UTF-16 (native endianness)
```

So basically, instead of UTF-16LE or UTF-16BE, we just write something like `UTF-16 (native endianness)` and do away with `Without depending on the system endianness;`.

File: include/simdutf/implementation.h [hunk 3]
```diff
@@ -91,6 +91,22 @@ simdutf_warn_unused bool validate_ascii(const char *buf, size_t len) noexcept;
  */
 simdutf_warn_unused result validate_ascii_with_errors(const char *buf, size_t len) noexcept;
 
+/**
+ * Without depending on the system endianness;
+ * Validate the UTF-16 string. This function may be best when you expect
+ * the input to be almost always valid. Otherwise, consider using
+ * validate_utf16le_with_errors.
```
Comment (lemire) [inline include/simdutf/implementation.h:98]: should be validate_utf16_with_errors.

File: include/simdutf/implementation.h [hunk 4]
```diff
@@ -91,6 +91,22 @@ simdutf_warn_unused bool validate_ascii(const char *buf, size_t len) noexcept;
  */
 simdutf_warn_unused result validate_ascii_with_errors(const char *buf, size_t len) noexcept;
 
+/**
+ * Without depending on the system endianness;
```
Comment (lemire) [inline include/simdutf/implementation.h:95]: Could be "Using native endianness;"

File: include/simdutf/implementation.h [hunk 5]
```diff
@@ -121,6 +137,21 @@ simdutf_warn_unused bool validate_utf16le(const char16_t *buf, size_t len) noexc
  */
 simdutf_warn_unused bool validate_utf16be(const char16_t *buf, size_t len) noexcept;
 
+/**
+ * Without depending on the system endianness;
+ * Validate the UTF-16 string and stop on error. It might be faster than
+ * validate_utf16le when an error is expected to occur early.
```
Comment (lemire) [inline include/simdutf/implementation.h:143]: should be validate_utf16

File: include/simdutf/implementation.h [hunk 6]
```diff
@@ -91,6 +91,22 @@ simdutf_warn_unused bool validate_ascii(const char *buf, size_t len) noexcept;
...
+ * the input to be almost always valid. Otherwise, consider using
+ * validate_utf16le_with_errors.
+ *
+ * Overridden by each implementation.
+ *
+ * This function is not BOM-aware.
+ *
+ * @param buf the UTF-16 string to validate.
+ * @param len the length of the string in number of 2-byte words (char16_t).
+ * @return true if and only if the string is valid UTF-16LE.
```
Comment (lemire) [inline include/simdutf/implementation.h:106]: UTF-16LE-> valid UTF-16 using native endianness

**Comment by anonrig** [comment]
Done @lemire. If everything ok, I'll update the readme.

**Comment by lemire** [comment]
@anonrig This is good. See two comments I have added. If you update the README, I will merge and issue a commit soon.

For the README, you can just copy-paste. For now, we don't have a fancy documentation in the README.

File: src/implementation.cpp [hunk 7]
```diff
@@ -651,6 +679,13 @@ simdutf_warn_unused bool validate_utf32(const char32_t * buf, size_t len) noexce
 simdutf_warn_unused result validate_utf32_with_errors(const char32_t * buf, size_t len) noexcept {
   return active_implementation->validate_utf32_with_errors(buf, len);
 }
+simdutf_warn_unused size_t convert_valid_utf8_to_utf16(const char * input, size_t length, char16_t* utf16_buffer) noexcept {
+  #if SIMDUTF_IS_BIG_ENDIAN
+  return convert_valid_utf8_to_utf16le(input, length, utf16_buffer);
```
Comment (anonrig) [inline src/implementation.cpp:684]: This is wrong.

**Comment by anonrig** [comment]
I've updated the readme, we can merge this @lemire

**Comment by lemire** [comment]
Release upcoming.

**Pull request merged**